src/server/server_family.cc

```diff
@@ -522,6 +516,247 @@ bool IsReplicatingNoOne(string_view host, string_view port) {
   return absl::EqualsIgnoreCase(host, "no") && absl::EqualsIgnoreCase(port, "one");
 }
 
+struct SaveStagesInputs {
+  bool use_dfs_format_;
+  string_view basename_;
+  Transaction* trans_;
+  Service* service_;
+  atomic_bool* is_saving_;
+  FiberQueueThreadPool* fq_threadpool_;
+  shared_ptr<LastSaveInfo>* last_save_info_;
+  Mutex* save_mu_;
+  unique_ptr<cloud::AWS>* aws_;
+};
+
+struct SaveStagesController : public SaveStagesInputs {
+  SaveStagesController(SaveStagesInputs&& inputs) : SaveStagesInputs{move(inputs)} {
+    start_time_ = absl::Now();
+  }
+
+  ~SaveStagesController() {
+    service_->SwitchState(GlobalState::SAVING, GlobalState::ACTIVE);
+  }
+
+  GenericError Save() {
+    if (auto err = BuildFullPath(); err)
+      return err;
+
+    if (auto err = SwitchState(); err)
+      return err;
+
+    if (auto err = InitResources(); err)
+      return err;
+
+    // The stages below report errors to shared_err_
+    if (use_dfs_format_)
+      SaveDfs();
+    else
+      SaveRdb();
+
+    is_saving_->store(true, memory_order_relaxed);
+    RunStage(&SaveStagesController::SaveCb);
+    is_saving_->store(false, memory_order_relaxed);
+
+    RunStage(&SaveStagesController::CloseCb);
+
+    FinalizeFileMovement(bool(shared_err_));
+
+    if (!shared_err_)
+      UpdateSaveInfo();
+
+    return *shared_err_;
+  }
+
+ private:
+  // In the new version (.dfs) we store a file for every shard and one more summary file.
+  // Summary file is always last in snapshots array.
+  void SaveDfs() {
+    // Extend all filenames with -{sid} or -summary and append .dfs.tmp
+    const string_view ext = is_cloud_ ? ".dfs" : ".dfs.tmp";
+    ShardId sid = 0;
+    for (auto& [_, filename] : snapshots_) {
+      filename = full_path_;
+      if (sid < shard_set->size())
+        ExtendDfsFilenameWithShard(sid++, ext, &filename);
+      else
+        SetExtension("summary", ext, &filename);
+    }
+
+    // Save summary file.
+    SaveDfsSingle(nullptr);
+
+    // Save shard files.
+    auto cb = [this](Transaction* t, EngineShard* shard) {
+      SaveDfsSingle(shard);
+      return OpStatus::OK;
+    };
+    trans_->ScheduleSingleHop(std::move(cb));
+  }
+
+  // Start saving a dfs file on shard
+  void SaveDfsSingle(EngineShard* shard) {
+    auto& [snapshot, filename] = snapshots_[shard ? shard->shard_id() : shard_set->size()];
+
+    SaveMode mode = shard == nullptr ? SaveMode::SUMMARY : SaveMode::SINGLE_SHARD;
+    auto scripts = shard == nullptr ? GetLuaScripts() : StringVec{};
+
+    if (auto err = snapshot->Start(mode, filename, scripts); err) {
+      shared_err_ = err;
+      snapshot.reset();
+      return;
+    }
+
+    if (mode == SaveMode::SINGLE_SHARD)
+      snapshot->StartInShard(shard);
+  }
+
+  // Save a single rdb file
+  void SaveRdb() {
+    auto& [snapshot, _] = snapshots_.front();
+
+    if (!full_path_.has_extension())
+      full_path_ += ".rdb";
+    if (!is_cloud_)
+      full_path_ += ".tmp";
```

Comment:
I changed it to add the .tmp suffix only if its not saved on cloud. Previously it was added in any case, yet snapshots are not renamed for the cloud env. I assume this was a bug